cf/latest-snapshot/src/index.ts

```diff
@@ -7,31 +9,56 @@ function basename(path: String) {
 }
 
 // Directly fetch the data for the latest snapshot of a given chain (eg. calibnet or mainnet)
-async function get_latest(env: Env, chain: string): Promise<Response> {
+async function get_latest(req_headers: Headers, env: Env, chain: string): Promise<Response> {
 	const listed = await env.FOREST_ARCHIVE.list({ prefix: chain + "/latest/" });
 	let latest = listed.objects.at(-1);
 	if (latest == null) {
 		return new Response(`No latest snapshot found ${chain}`, {
 			status: 404,
 		});
 	} else {
-		// Should we support range queries?
-		const object = await env.FOREST_ARCHIVE.get(latest.key);
+		const object = await env.FOREST_ARCHIVE.get(latest.key, {
+			range: req_headers,
+			onlyIf: req_headers,
+		});
+
 		if (object === null) {
 			return new Response('No latest snapshot found', {
 				status: 404,
 			});
 		}
+
 		const headers = new Headers();
+
+		// If the client requested a range, then we need to set the Content-Range header.
+		// The `parseRange` returns an error code if the range is not satisfiable (or not specified),
+		// so we can handle the range request only if it returns an array.
+		let status = 200;
+		let range = parseRange(object.size, req_headers.get("range") || "");
+		if (Array.isArray(range)) {
+			status = 206; // Partial Content
+			range.forEach(function (r) {
+				headers.append("Content-Range", `${range.type} ${r.start}-${r.end}/${object.size}`);
+			})
+		}
+
 		object.writeHttpMetadata(headers);
 		headers.set('etag', object.httpEtag);
-		let encoded_name = encodeURIComponent(basename(object.key));
+		const encoded_name = encodeURIComponent(basename(object.key));
 		// Tell browsers and aria2c which filename to use. For 'wget', you have to use `--trust-server-names`.
-		headers.set('content-disposition', `attachment; filename*=UTF-8''${encoded_name}; filename="${encoded_name}"`);
+		headers.set('Content-Disposition', `attachment; filename*=UTF-8''${encoded_name}; filename="${encoded_name}"`);
 
-		return new Response(object.body, {
-			headers,
-		});
+		if ('body' in object) {
```

Comment:
When does `object` not have `body`?

Reply:
I have no clue under what conditions the `get` would return an `R2Object` (and not an `R2ObjectBody`). It is a possible output from this method, though, so it needs to be handled (otherwise, we would have a runtime error).

Reply:
It always returns both, doesn't it?

Reply:
The returned object implements both `R2Object` and `R2ObjectBody`.

Reply:
Does it implement _both_, though? Or does it implement _either_? I got warnings that `R2Object` does not have a `body` property.

Reply:
That is to say, I do believe it's a union (in C sense) where only one variant is valid.